web/quickfeed_service.go

```diff
@@ -563,17 +563,27 @@ func (s *QuickFeedService) GetOrganization(ctx context.Context, in *connect.Requ
 }
 
 // GetRepositories returns URL strings for repositories of given type for the given course.
-func (s *QuickFeedService) GetRepositories(ctx context.Context, in *connect.Request[qf.URLRequest]) (*connect.Response[qf.Repositories], error) {
+func (s *QuickFeedService) GetRepositories(ctx context.Context, in *connect.Request[qf.CourseRequest]) (*connect.Response[qf.Repositories], error) {
 	course, err := s.db.GetCourse(in.Msg.GetCourseID(), false)
 	if err != nil {
 		s.logger.Errorf("GetRepositories failed: course %d not found: %v", in.Msg.GetCourseID(), err)
 		return nil, connect.NewError(connect.CodeNotFound, errors.New("course not found"))
 	}
 	usrID := userID(ctx)
-	enrol, _ := s.db.GetEnrollmentByCourseAndUser(course.GetID(), usrID)
+	enrol, err := s.db.GetEnrollmentByCourseAndUser(course.GetID(), usrID)
+	if err != nil {
+		s.logger.Error("GetRepositories failed: enrollment for user %d and course %d not found", usrID, course.GetID())
+		return nil, connect.NewError(connect.CodeNotFound, errors.New("enrollment not found"))
+	}
 
 	urls := make(map[string]string)
-	for _, repoType := range in.Msg.GetRepoTypes() {
+	for _, repoType := range repoTypes(enrol) {
+		if repoType == qf.Repository_GROUP && enrol.GroupID == 0 {
+			// If groupID for a group repository type is 0, getRepo() returns the first group repository for the course
+			// because database query looks like this: Repository{type: GROUP, OrganizationID: course.OrgID} which mathes
+			// to any group repository for this course. To prevent this, ignore such cases instead of calling getRepo().
+			continue
+		}
 		var id uint64
 		switch repoType {
 		case qf.Repository_USER:
```

Comment:
Referring to Line 596, if you implement my suggestion from [issue #931 (comment)](https://github.com/quickfeed/quickfeed/issues/941#issuecomment-1332652355), you can use:
```go
urls[repoType] = repo.GetHTMLURL()
```
(I think)...